lib/srv/db/proxyserver.go

```diff
@@ -85,12 +88,51 @@ type ProxyServerConfig struct {
 	Clock clockwork.Clock
 	// ServerID is the ID of the audit log server.
 	ServerID string
-	// Shuffle allows to override shuffle logic in tests.
-	Shuffle func([]types.DatabaseServer) []types.DatabaseServer
 	// LockWatcher is a lock watcher.
 	LockWatcher *services.LockWatcher
 }
 
+// ShuffleFunc defines a function that shuffles a list of database servers.
+type ShuffleFunc func([]types.DatabaseServer) []types.DatabaseServer
+
+// ShuffleRandom is a ShuffleFunc that randomizes the order of database servers.
+// Used to provide load balancing behavior when proxying to multiple agents.
+func ShuffleRandom(servers []types.DatabaseServer) []types.DatabaseServer {
+	rand.New(rand.NewSource(time.Now().UnixNano())).Shuffle(
+		len(servers), func(i, j int) {
+			servers[i], servers[j] = servers[j], servers[i]
+		})
+	return servers
+}
+
+// ShuffleSort is a ShuffleFunc that sorts database servers by name and host ID.
+// Used to provide predictable behavior in tests.
+func ShuffleSort(servers []types.DatabaseServer) []types.DatabaseServer {
+	sort.Sort(types.DatabaseServers(servers))
+	return servers
+}
+
+var (
+	// mu protects the shuffleFunc global access.
+	mu sync.RWMutex
+	// shuffleFunc provides shuffle behavior for multiple database agents.
+	shuffleFunc ShuffleFunc = ShuffleRandom
```

Comment:
I'm not big fun of global variables but I assume that this logic is hard to test without this. Though wonder if the db servers order can be enforced in integration tests by settings specific time point in clock used by proxy db service:
```
clock := NewFakeClockAt(time.Date(2021, time.April, 4, 0, 0, 0, 0, time.UTC))
pack := setupDatabaseTest(t, withClock(clock))
```
 https://github.com/gravitational/teleport/pull/10641/files#diff-3ed9752f98ccf36539f32e345ea5cb26a13d71552d6b9a9be5117308d26b00cbL119

Reply:
Yeah I thought about the same actually but didn't know how to ensure the stable shuffle order even with the frozen time. I think in this case shuffle will always produce the same order but there's no way of telling _which_ order. So while I'm not a big fan of this either, this felt like the most reliable way to ensure guaranteed order.

Reply:
Technically we could always shuffle and in tests add an extra pass to sort them.

Reply:
If you mean adding some sort of "test mode", I was trying to avoid that. I generally try not to make "if isTestMode()" switches in the code.

Reply:
I was about to write NVM, but you were faster. Initially I thought about passing extra sort function in test, but that introduces the same issue that we have already. 
I agree on the "test mode" too.